camp/abolotski/smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -0,0 +1,968 @@
+    backRightModule.set(
+      states[3].speedMetersPerSecond / MAX_VELOCITY_METERS_PER_SECOND * MAX_VOLTAGE,
+      states[3].angle.getRadians());
+  }
+
+  /**
+   * a method to print relevant state of the navx
+   */
+  public void printState(){
+    System.out.println("**** NavX State ****");
+    System.out.println("Quaternion ------>" + this.getQuaterion());
+    System.out.println("Roll, Pitch, Yaw ------>" + this.getEulerAngle());
+    System.out.println("Is the robot level? -------->" + this.isLevel());
+    System.out.println("SteepestAscent --->" + VectorUtils.getAngleOfSteepestAscent(getEulerAngle()));
+  }
+
+  /**
+   * A method to set the current position of the robot
+   * @param updatedPosition - the new position of the robot
+   */
+  public void setRobotPosition(Pose2d updatedPosition)
+  {
+    //TOD this didn't work to fix the problem of the robot not getting correct initial position.  
+    // try to debug why not.  
+    try{
+      theLock.lock();
+      // initialize the odometry goo
+      currentPosition = updatedPosition;
+      this.initializeSwerveOdometry(currentPosition);
+    }
+    finally {
+      theLock.unlock();
+    }
+  }
+
+  /**
+   * A method to set the swerve drive mode
+   * @param swerveDriveMode
+   */
+  public void setSwerveDriveMode(SwerveDriveMode swerveDriveMode) {
+    this.swerveDriveMode = swerveDriveMode;
+    // if (swerveDriveMode == SwerveDriveMode.IMMOVABLE_STANCE) DO NOT set chassis speeds to 0.0 here.
+    // we need to allow the robot to decel properly so that it doesn't tip.  
+  }
+
+
+  /**
+   * A method to set the swerve rotation mode
+   * @param swerveDriveRotationMode
+   */
+  public void setSwerveDriveCenterOfRotation(SwerveDriveCenterOfRotation swerveDriveCenterOfRotation) {
+    this.swerveDriveCenterOfRotation = swerveDriveCenterOfRotation;
+  }
+
+  /**
+   * sets the Yaw to a specific angle
+   * @param offsetDegrees
+   */
+  public void setYaw(double offsetDegrees) {
+    this.zeroGyroscope();
+    this.yawOffsetDegrees = offsetDegrees;
+  }
+
+  /**
+   * Sets the gyroscope angle to zero. This can be used to set the direction the robot is currently facing to the
+   * 'forwards' direction.
+   */
+  public void zeroGyroscope() {
+    if(swerveNavx.isCalibrating()){
+      // From the NavX Docs: This method has no effect if the sensor is currently calibrating
+      System.out.println("WARNING: Gyro is calibrating. Zeroing gyro has no effect while it is calibrating.");
+    }
+    swerveNavx.zeroYaw();
+    this.yawOffsetDegrees = 0.0;
+  }
+
+  /**
+   * A method to zero the current position
+   */
+  public void zeroRobotPosition()
+  {
+    this.setRobotPosition(new Pose2d(0,0,Rotation2d.fromDegrees(0)));
+  }
+
+  /**
+   * a method that updates the robot position with vision
+   */
+  private void addVisionMeasurement(Pose2d RobotPos, double Timestamp){
+    VisionMeasurement visionMeasurement = cameraSubsystem.getVisionPosition();
```

Comment:
This does not have the change I merged to your branch over in ChargedUp2023

Reply:
![image](https://github.com/Team4682CyBears/Crescendo2024/assets/56608471/ef1fa7fa-6e85-4091-9123-4e9c7bd736f8)

Reply:
fixed